### PR TITLE
Multi-wallet: Enable users to connect with multiple wallets

### PR DESCRIPTION
**Pull request by kaitoo1**
MULTI WALLET IS HERE

### CODE REVIEW

File: src/components/ManageWallets/ManageWallets.tsx [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+type Props = {
+  newAddress?: string;
+};
+
+function ManageWallets({ newAddress }: Props) {
+  const addresses = useAuthenticatedUserAddresses();
+  const fetcher = useFetcher();
+
+  const [localAddresses, setLocalAddresses] = useState(addresses);
```
Comment (Robinnnnn) [inline src/components/ManageWallets/ManageWallets.tsx:23]: handle this with optimistic update / mutation

File: src/components/ManageWallets/ManageWallets.tsx [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+  const removeUserAddressInternally = useCallback((index: number) => {
+    localAddresses.splice(index, 1);
+    setLocalAddresses([...localAddresses]);
+  }, [localAddresses]);
+
+  useEffect(() => {
+    setLocalAddresses([...addresses]);
+  }, [addresses]);
+
+  const userSigninAddress = useMemo(() => getLocalStorageItem<string>(USER_SIGNIN_ADDRESS_LOCAL_STORAGE_KEY), []) ?? '';
```
Comment (Robinnnnn) [inline src/components/ManageWallets/ManageWallets.tsx:41]: usePersistedState

File: src/components/ManageWallets/ManageWalletsRow.tsx [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+  setErrorMessage: (message: string) => void;
+};
+
+function ManageWalletsRow({ index, address, userSigninAddress, fetcher, onDisconnect, setErrorMessage }: Props) {
+  const handleDisconnectClick = useCallback(async () => {
+    try {
+      setErrorMessage('');
+      await removeUserAddress({ addresses: [address] }, fetcher).then(() => {
+        onDisconnect(index);
+      });
```
Comment (Robinnnnn) [inline src/components/ManageWallets/ManageWalletsRow.tsx:26]: don't need `.then` if in async/await

```ts
await removeUserAddress({ addresses: [address] }, fetcher)
onDisconect(index)
```

File: src/components/WalletSelector/AddWalletPending.tsx [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+        } else {
+          await attemptAddWallet(account.toLowerCase(), signer);
+        }
+      }
+    }
+
+    void authenticate();
+  }, [account, signer, authenticatedUserAddresses, attemptAddWallet, isMetamask]);
+
+  if (pendingState === ADDRESS_ALREADY_CONNECTED) {
```
Comment (Robinnnnn) [inline src/components/WalletSelector/AddWalletPending.tsx:93]: add `&& account` so we don't have to optional chain later

File: src/components/WalletSelector/AddWalletPending.tsx [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+  if (pendingState === PROMPT_SIGNATURE) {
+    return (
+      <div>
+        <StyledTitleMedium>Connect with {userFriendlyWalletName}</StyledTitleMedium>
+        <BodyRegular color={colors.gray50}>Sign the message with your wallet.</BodyRegular>
+      </div>
+    );
+  }
+
+  return (
```
Comment (Robinnnnn) [inline src/components/WalletSelector/AddWalletPending.tsx:130]: add comment above this block to explain this is `pendingState === INITIAL`

File: src/components/WalletSelector/authRequestUtils.ts [hunk 6]
```diff
@@ -24,6 +23,30 @@ type AuthPipelineProps = {
...
+  address,
+  signer,
+  fetcher,
+  connector,
+}: AuthPipelineProps): Promise<AddWalletResult> {
+  const { nonce, user_exists: userExists } = await fetchNonce(address, fetcher);
+
+  if (userExists) {
+    throw { code: 'EXISTING_USER' } as Web3Error;
+  }
```
Comment (Robinnnnn) [inline src/components/WalletSelector/authRequestUtils.ts:40]: add comment explaining what's happening?

File: src/components/WalletSelector/AuthenticateWalletPending.tsx [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+
+  const attemptAuthentication = useCallback(async (address: string, signer: JsonRpcSigner) => {
+    setPendingState(PROMPT_SIGNATURE);
+    const { jwt, userId } = await initializeAuthPipeline({
+      address,
+      signer,
+      fetcher,
+      connector: pendingWallet,
+    });
+    Mixpanel.trackConnectWallet(pendingWalletName);
```
[the remark this answers is not in the document]
Comment (kaitoo1) [reply src/components/WalletSelector/AuthenticateWalletPending.tsx:44]: good idea, done ✅

File: src/components/WalletSelector/AddWalletPending.tsx (hunk 5 above)
Comment (kaitoo1) [reply src/components/WalletSelector/AddWalletPending.tsx:130]: done ✅

File: src/components/WalletSelector/AddWalletPending.tsx (hunk 4 above)
Comment (kaitoo1) [reply src/components/WalletSelector/AddWalletPending.tsx:93]: done ✅

File: src/components/ManageWallets/ManageWalletsRow.tsx (hunk 3 above)
Comment (kaitoo1) [reply src/components/ManageWallets/ManageWalletsRow.tsx:26]: done ✅

File: src/components/WalletSelector/authRequestUtils.ts (hunk 6 above)
Comment (kaitoo1) [reply src/components/WalletSelector/authRequestUtils.ts:40]: done ✅

File: src/components/ManageWallets/ManageWallets.tsx (hunk 1 above)
Comment (kaitoo1) [reply src/components/ManageWallets/ManageWallets.tsx]: done ✅

File: src/components/ManageWallets/ManageWallets.tsx (hunk 2 above)
Comment (kaitoo1) [reply src/components/ManageWallets/ManageWallets.tsx]: done ✅

**Pull request merged**